drivers/misc/nxp_rtxxx_adsp_ctl/nxp_rtxxx_adsp_ctl.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_rtxxx_adsp_ctl
+
+#include <zephyr/drivers/misc/nxp_rtxxx_adsp_ctl/nxp_rtxxx_adsp_ctl.h>
+#include <zephyr/dt-bindings/misc/nxp_rtxxx_adsp_ctl.h>
+
+#include <fsl_device_registers.h>
+#include <fsl_dsp.h>
+#include <fsl_clock.h>
+
+#include <errno.h>
+#include <zephyr/init.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+
+struct nxp_rtxxx_adsp_ctl_region {
+    void* base;
+    int32_t length;
+};
+
+struct nxp_rtxxx_adsp_ctl_config {
+    SYSCTL0_Type* sysctl;
+    struct nxp_rtxxx_adsp_ctl_region regions[NXP_RTXXX_ADSP_REGION_MAX];
+};
+
+static void adsp_ctl_enable(const struct device *dev)
+{
+    SYSCTL0_Type *sysctl = ((struct nxp_rtxxx_adsp_ctl_config*)dev->config)->sysctl;
+    sysctl->DSPSTALL = 0;
+}
+
+static void adsp_ctl_disable(const struct device *dev)
+{
+    SYSCTL0_Type *sysctl = ((struct nxp_rtxxx_adsp_ctl_config*)dev->config)->sysctl;
+    sysctl->DSPSTALL = 1;
+}
+
+static int adsp_ctl_load_section(
+    const struct device *dev, 
+    const struct nxp_rtxxx_adsp_ctl_image* image, 
+    enum nxp_rtxxx_adsp_ctl_section_type section
+)
+{
+    if(section >= NXP_RTXXX_ADSP_REGION_MAX)
+        return -EINVAL;
+
+    struct nxp_rtxxx_adsp_ctl_config *cfg = (struct nxp_rtxxx_adsp_ctl_config*)dev->config;
+    if(cfg->regions[section].base == NULL)
+        return -EINVAL;
+
+    if(image->length > cfg->regions[section].length)
+        return -ENOMEM;
+
+    /* Custom memcpy implementation is needed because the DSP TCMs can be accessed only by 32 bits. */
+
+    uint32_t *src = image->base;
+    uint32_t *dst = cfg->regions[section].base;
+    uint32_t size = image->length;
+
+    while(size > 0)
```

Comment:
If to use "for" instead of "while", the code inside { } will have only one line.